### ISSUE TITLE
Strange tab completion in REPL

### ISSUE THREAD

**Issue by bkamins** [JuliaLang/julia#44841]
In some cases tab completion uses property names, and in other field names of an object. Example:
```
julia> using DataFrames

julia> gdf = groupby(DataFrame(a=1:3, b=4:6), :a);

julia> gdf[1].

colindex parent    rows
julia> x = gdf[1]
1×2 SubDataFrame
 Row │ a      b
     │ Int64  Int64
─────┼──────────────
   1 │     1      4

julia> x.

a b
```
What is the reason for this?

**Comment by bkamins** [comment]
Note that tab completion for `gdf[1].` uses field names of the object. While after we assign this object to other variable with `x = gdf[1]` then property names are used by tab completion.

**Comment by Moelf** [comment]
```julia
gdf[1].
```

I think it never worked:
```julia
julia> a = Any[rand(ComplexF16, 2)...];

julia> 

julia> x = a[1]
Float16(0.748) + Float16(0.1914)im

julia> x.
im re
julia> a[1].
# tab gives you nothing
```

**Comment by aviatesk** [comment]
The field completion for `x[y].` kind of pattern uses type inference to get a type of retrieved object, and type inference couldn't infer the type `ComplexF16` for your case (just because the parent object is typed as `Vector{Any}`).

**Comment by bkamins** [comment]
> I don't think it's a bug

Indeed https://docs.julialang.org/en/v1/stdlib/REPL/#Tab-completion explains that "fields" are expanded using tab completion.

However there are two problems with this (and that is why I would say it is a bug):
* tab-completing fields is incorrect for types that do not expose fields but have custom properties (this is the whole point of having the distinction between them in Julia);
* for objects directly bound to a variable (as I have shown above) tab completion lists properties not fields.

In summary - I agree that it has always worked like this. It is just that some user has recently reported it as confusing and I agree with this observation. I propose to discuss how or if this should be fixed.

**Comment by Moelf** [comment]
I think the fix is for the first case to return nothing?

**Comment by bkamins** [comment]
My opinion: first best would be to return properties always (and update the documentation accordingly). Second best would be to return properties when possible, and in other cases do not autocomplete at all (and update the documentation accordingly).

The problem is that now we sometimes return properties and sometimes return fields which IMO is confusing the users.

**Comment by Moelf** [comment]
First is impossible due to unknown cost or side effects of indexing/get property, we can make assumptions, but I guess it's a triage topic

**Comment by KristofferC** [comment]
Fields vs properties could also be colored slightly differently in the repl perhaps.

**Comment by nalimilan** [comment]
BTW, the docs don't say anywhere that `getproperty` should call `getfield` to allow accessing private fields. So in the example above, typing `gdf[1].colindex` as autocompletion suggests throws an error, because `getproperty(::GroupedDataFrame, property)` doesn't allow accessing private fields (this is safer as their names could conflict with column names). This makes the suggestion totally wrong.

**Comment by LilithHafner** [comment]
The OP no longer reproduces (nothing tab completes for `gdf[1].[tab]`), and when the compiler knows that the construction is cheap, property names are tab completed.
```julia
julia> struct T
           field1::Int
           field2::Int
       end

julia> Base.propertynames(::T) = (:property1,:property2)

julia> T(5,3).property
property1
property2
```
